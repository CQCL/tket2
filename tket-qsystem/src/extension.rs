//! This module defines the Hugr extensions used by tket-qsystem.
pub mod futures;
pub mod qsystem;
pub mod random;
pub mod result;
pub mod utils;
pub mod wasm;
