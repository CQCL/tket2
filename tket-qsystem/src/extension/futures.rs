//! This module defines the Hugr extension used to represent futures.
//!
//! `Future<t>` is a linear type representing a value that will be available in
//! the future.  It can be consumed by `Read`, returning a `t`.  It can be
//! duplicated by `Dup`, and discarded with `Free`.
use std::sync::{Arc, Weak};

use hugr::{
    builder::{BuildError, Dataflow},
    extension::{
        simple_op::{
            try_from_name, HasConcrete, HasDef, MakeExtensionOp, MakeOpDef, MakeRegisteredOp,
            OpLoadError,
        },
        ExtensionBuildError, ExtensionId, OpDef, SignatureError, SignatureFunc, TypeDef, Version,
    },
    ops::{custom::ExtensionOp, OpType},
    types::{type_param::TypeParam, CustomType, PolyFuncType, Signature, Type, TypeArg, TypeBound},
    Extension, Wire,
};
use lazy_static::lazy_static;
use serde::{Deserialize, Serialize};
use smol_str::SmolStr;
use strum::{EnumIter, EnumString, IntoStaticStr};

/// The ID of the `tket.futures` extension.
pub const EXTENSION_ID: ExtensionId = ExtensionId::new_unchecked("tket.futures");
/// The "tket.futures" extension version
pub const EXTENSION_VERSION: Version = Version::new(0, 2, 0);

lazy_static! {
    /// The "tket.futures" extension.
    pub static ref EXTENSION: Arc<Extension>  = {
        Extension::new_arc(EXTENSION_ID, EXTENSION_VERSION, |ext, ext_ref| {
            let _ = add_future_type_def(ext, ext_ref.clone()).unwrap();

            FutureOpDef::load_all_ops( ext, ext_ref).unwrap();
        })
    };

    /// The name of the `Future` type.
    pub static ref FUTURE_TYPE_NAME: SmolStr = SmolStr::new_inline("Future");
}

fn add_future_type_def(
    ext: &mut Extension,
    extension_ref: Weak<Extension>,
) -> Result<&TypeDef, ExtensionBuildError> {
    ext.add_type(
        FUTURE_TYPE_NAME.to_owned(),
        vec![TypeBound::Linear.into()],
        "A value that is computed asynchronously".into(),
        TypeBound::Linear.into(),
        &extension_ref,
    )
}

/// Returns a `Future<t>` [CustomType].
pub fn future_custom_type(t: Type, extension_ref: &Weak<Extension>) -> CustomType {
    CustomType::new(
        FUTURE_TYPE_NAME.to_owned(),
        vec![t.into()],
        EXTENSION_ID,
        TypeBound::Linear,
        extension_ref,
    )
}

/// Returns a `Future<t>` [Type].
pub fn future_type(t: Type) -> Type {
    future_custom_type(t, &Arc::downgrade(&EXTENSION)).into()
}

#[derive(
    Clone,
    Copy,
    Debug,
    Serialize,
    Deserialize,
    Hash,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    EnumIter,
    IntoStaticStr,
    EnumString,
)]
#[allow(missing_docs)]
#[non_exhaustive]
/// Simple enum of "tket.futures" operations.
pub enum FutureOpDef {
    Read,
    Dup,
    Free,
}

impl MakeOpDef for FutureOpDef {
    fn opdef_id(&self) -> hugr::ops::OpName {
        <&'static str>::from(self).into()
    }

    fn init_signature(&self, extension_ref: &Weak<Extension>) -> SignatureFunc {
        let t_param = TypeParam::from(TypeBound::Linear);
        let t_type = Type::new_var_use(0, TypeBound::Linear);
        let future_type = Type::new_extension(future_custom_type(t_type.clone(), extension_ref));
        match self {
            FutureOpDef::Read => {
                PolyFuncType::new([t_param], Signature::new(future_type, t_type)).into()
            }
            FutureOpDef::Dup => PolyFuncType::new(
                [t_param],
                Signature::new(future_type.clone(), vec![future_type.clone(), future_type]),
            )
            .into(),
            FutureOpDef::Free => {
                PolyFuncType::new([t_param], Signature::new(future_type.clone(), vec![])).into()
            }
        }
    }

    fn extension(&self) -> ExtensionId {
        EXTENSION_ID
    }

    fn from_def(op_def: &OpDef) -> Result<Self, hugr::extension::simple_op::OpLoadError> {
        try_from_name(op_def.name(), op_def.extension_id())
    }

    fn description(&self) -> String {
        match self {
            FutureOpDef::Read => "Read a value from a Future, consuming it".into(),
            FutureOpDef::Dup => {
                "Duplicate a Future. The original Future is consumed and two Futures are returned"
                    .into()
            }
            FutureOpDef::Free => "Consume a future without reading it.".into(),
        }
    }

    fn extension_ref(&self) -> Weak<Extension> {
        Arc::downgrade(&EXTENSION)
    }
}

impl HasConcrete for FutureOpDef {
    type Concrete = FutureOp;

    fn instantiate(&self, type_args: &[TypeArg]) -> Result<Self::Concrete, OpLoadError> {
        match type_args {
            [TypeArg::Runtime(ty)] => Ok(FutureOp {
                op: *self,
                typ: ty.clone(),
            }),
            _ => Err(SignatureError::InvalidTypeArgs.into()),
        }
    }
}

impl From<&FutureOp> for &'static str {
    fn from(value: &FutureOp) -> Self {
        value.op.into()
    }
}

/// Concrete "tket.futures" operations with type set.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct FutureOp {
    /// The `FutureOpDef` that defines this operation.
    pub op: FutureOpDef,
    /// The inner type of the `Future` this op acts on.
    pub typ: Type,
}

impl MakeExtensionOp for FutureOp {
    fn op_id(&self) -> hugr::ops::OpName {
        self.op.opdef_id()
    }

    fn from_extension_op(ext_op: &ExtensionOp) -> Result<Self, OpLoadError>
    where
        Self: Sized,
    {
        FutureOpDef::from_def(ext_op.def())?.instantiate(ext_op.args())
    }

    fn type_args(&self) -> Vec<hugr::types::TypeArg> {
        vec![self.typ.clone().into()]
    }
}

impl MakeRegisteredOp for FutureOp {
    fn extension_id(&self) -> ExtensionId {
        EXTENSION_ID
    }

    fn extension_ref(&self) -> Weak<Extension> {
        Arc::downgrade(&EXTENSION)
    }
}

impl HasDef for FutureOp {
    type Def = FutureOpDef;
}

impl TryFrom<&OpType> for FutureOpDef {
    type Error = OpLoadError;

    fn try_from(value: &OpType) -> Result<Self, Self::Error> {
        Self::from_op(
            value
                .as_extension_op()
                .ok_or(OpLoadError::NotMember(value.to_string()))?,
        )
    }
}

/// An extension trait for [Dataflow] providing methods to add "tket.futures"
/// operations.
pub trait FutureOpBuilder: Dataflow {
    /// Add a "tket.futures.Read" op.
    fn add_read(&mut self, lifted: Wire, typ: Type) -> Result<[Wire; 1], BuildError> {
        Ok(self
            .add_dataflow_op(
                FutureOp {
                    op: FutureOpDef::Read,
                    typ,
                },
                [lifted],
            )?
            .outputs_arr())
    }

    /// Add a "tket.futures.Dup" op.
    fn add_dup(&mut self, lifted: Wire, typ: Type) -> Result<[Wire; 2], BuildError> {
        Ok(self
            .add_dataflow_op(
                FutureOp {
                    op: FutureOpDef::Dup,
                    typ,
                },
                [lifted],
            )?
            .outputs_arr())
    }

    /// Add a "tket.futures.Free" op.
    fn add_free(&mut self, lifted: Wire, typ: Type) -> Result<(), BuildError> {
        let op = self.add_dataflow_op(
            FutureOp {
                op: FutureOpDef::Free,
                typ,
            },
            [lifted],
        )?;
        assert!(op.outputs().len() == 0);
        Ok(())
    }
}

impl<D: Dataflow> FutureOpBuilder for D {}

#[cfg(test)]
pub(crate) mod test {

    use hugr::builder::{Dataflow, DataflowHugr, FunctionBuilder};
    use hugr::HugrView;
    use std::sync::Arc;
    use strum::IntoEnumIterator;

    use super::*;

    fn get_opdef(op: FutureOpDef) -> Option<&'static Arc<OpDef>> {
        EXTENSION.get_op(&op.opdef_id())
    }

    #[test]
    fn create_extension() {
        assert_eq!(EXTENSION.name(), &EXTENSION_ID);

        for o in FutureOpDef::iter() {
            assert_eq!(FutureOpDef::from_def(get_opdef(o).unwrap()), Ok(o));
        }
    }

    #[test]
    fn future_op_from_def() {
        let typ = Type::UNIT;

        assert_eq!(
            FutureOp {
                op: FutureOpDef::Free,
                typ: typ.clone()
            },
            FutureOpDef::Free.instantiate(&[typ.into()]).unwrap()
        )
    }

    #[test]
    fn circuit() {
        let t_param = TypeParam::from(TypeBound::Linear);
        let t = Type::new_var_use(0, TypeBound::Linear);
        let future_type = future_type(t.clone());

        let hugr = {
            let mut func_builder = FunctionBuilder::new(
                "circuit",
                PolyFuncType::new(vec![t_param], Signature::new(future_type, t.clone())),
            )
            .unwrap();
            let [future_w] = func_builder.input_wires_arr();
            let [future_w, lazy_dup_w] = func_builder.add_dup(future_w, t.clone()).unwrap();
            func_builder.add_free(future_w, t.clone()).unwrap();
            let [t_w] = func_builder.add_read(lazy_dup_w, t).unwrap();
            func_builder.finish_hugr_with_outputs([t_w]).unwrap()
        };
        hugr.validate().unwrap();
    }
}
