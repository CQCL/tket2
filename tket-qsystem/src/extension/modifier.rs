use hugr::{Hugr, HugrView};

use crate::QSystemPassError;
use tket::rich_circuit::modifier_resolver::resolve_modifier_with_entrypoints;

#[allow(missing_docs)]
#[derive(Default)]
pub struct ModifierResolverPass;

impl ModifierResolverPass {
    /// Run ModifierResolverPass.
    pub fn run(self, hugr: &mut Hugr) -> Result<(), QSystemPassError> {
        resolve_modifier_with_entrypoints(hugr, [hugr.entrypoint()]).unwrap();
        return Ok(());
    }
}
