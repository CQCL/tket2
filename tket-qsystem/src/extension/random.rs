//! This module defines the "tket.qsystem.random" extension, which includes
//! random number generation (RNG) functions available for Quantinuum systems.

use std::sync::{Arc, Weak};

use derive_more::derive::Display;
use hugr::{
    builder::{BuildError, Dataflow},
    extension::{
        prelude::{option_type, UnwrapBuilder},
        simple_op::{try_from_name, MakeOpDef, MakeRegisteredOp},
        ExtensionBuildError, ExtensionId, ExtensionRegistry, OpDef, SignatureFunc, TypeDefBound,
        Version, PRELUDE,
    },
    std_extensions::arithmetic::{float_types::float64_type, int_types::int_type},
    types::{CustomType, Signature, Type, TypeBound},
    Extension, Wire,
};
use lazy_static::lazy_static;
use smol_str::SmolStr;
use strum::{EnumIter, EnumString, IntoStaticStr};

/// The extension ID for the RNG extension.
pub const EXTENSION_ID: ExtensionId = ExtensionId::new_unchecked("tket.qsystem.random");
/// The version of the "tket.qsystem.random" extension.
pub const EXTENSION_VERSION: Version = Version::new(0, 2, 0);

lazy_static! {
    /// The "tket.qsystem.random" extension.
    pub static ref EXTENSION: Arc<Extension> = {
         Extension::new_arc(EXTENSION_ID, EXTENSION_VERSION, |ext, ext_ref| {
            add_random_type_defs(ext, ext_ref).unwrap();
            RandomOp::load_all_ops( ext, ext_ref).unwrap();
        })
    };

    /// Extension registry including the "tket.qsystem.random" extension and
    /// dependencies.
    pub static ref REGISTRY: ExtensionRegistry = ExtensionRegistry::new([
        EXTENSION.to_owned(),
        PRELUDE.to_owned(),
    ]);

    /// The name of the `tket.qsystem.random.context` type.
    pub static ref CONTEXT_TYPE_NAME: SmolStr = SmolStr::new_inline("context");
}

fn add_random_type_defs(
    extension: &mut Extension,
    extension_ref: &Weak<Extension>,
) -> Result<(), ExtensionBuildError> {
    extension.add_type(
        CONTEXT_TYPE_NAME.to_owned(),
        vec![],
        "The linear RNG context type".into(),
        TypeDefBound::any(),
        extension_ref,
    )?;
    Ok(())
}

/// An enum for RNG extension types.
pub enum RandomType {
    /// The linear RNG context type.
    RNGContext,
}

impl RandomType {
    fn get_type(&self, extension_ref: &Weak<Extension>) -> Type {
        match self {
            Self::RNGContext { .. } => CustomType::new(
                CONTEXT_TYPE_NAME.to_owned(),
                vec![],
                EXTENSION_ID,
                TypeBound::Linear,
                extension_ref,
            ),
        }
        .into()
    }
}

#[derive(
    Clone,
    Copy,
    Debug,
    serde::Serialize,
    serde::Deserialize,
    Hash,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    EnumIter,
    IntoStaticStr,
    EnumString,
    Display,
)]
/// The operations provided by the random extension.
pub enum RandomOp {
    /// `fn random_int(RNGContext) -> (u32, RNGContext)`
    RandomInt,
    /// `fn random_float(RNGContext) -> (f32, RNGContext)`
    RandomFloat,
    /// `fn random_int_bounded(RNGContext, bound: u32) -> (u32, RNGContext)`
    RandomIntBounded,
    /// `fn new_rng_context(seed: u64) -> Option<RNGContext>` // return None on second call
    NewRNGContext,
    /// `fn delete_rng_context(RNGContext) -> ()`
    DeleteRNGContext,
}

impl MakeOpDef for RandomOp {
    fn opdef_id(&self) -> hugr::ops::OpName {
        <&'static str>::from(self).into()
    }

    fn init_signature(&self, extension_ref: &std::sync::Weak<Extension>) -> SignatureFunc {
        match self {
            RandomOp::RandomInt => Signature::new(
                vec![RandomType::RNGContext.get_type(extension_ref)],
                vec![int_type(5), RandomType::RNGContext.get_type(extension_ref)],
            ),
            RandomOp::RandomFloat => Signature::new(
                vec![RandomType::RNGContext.get_type(extension_ref)],
                vec![
                    float64_type(),
                    RandomType::RNGContext.get_type(extension_ref),
                ],
            ),
            RandomOp::RandomIntBounded => Signature::new(
                vec![RandomType::RNGContext.get_type(extension_ref), int_type(5)],
                vec![int_type(5), RandomType::RNGContext.get_type(extension_ref)],
            ),
            RandomOp::NewRNGContext => Signature::new(
                vec![int_type(6)],
                Type::from(option_type(RandomType::RNGContext.get_type(extension_ref))),
            ),
            RandomOp::DeleteRNGContext => {
                Signature::new(vec![RandomType::RNGContext.get_type(extension_ref)], vec![])
            }
        }
        .into()
    }

    fn from_def(op_def: &OpDef) -> Result<Self, hugr::extension::simple_op::OpLoadError> {
        try_from_name(op_def.name(), op_def.extension_id())
    }

    fn extension(&self) -> ExtensionId {
        EXTENSION_ID
    }

    fn extension_ref(&self) -> std::sync::Weak<Extension> {
        Arc::downgrade(&EXTENSION)
    }

    fn description(&self) -> String {
        match self {
            RandomOp::RandomInt => "Generate a random 32-bit unsigned integer.",
            RandomOp::RandomFloat => "Generate a random floating point value in the range [0,1).",
            RandomOp::RandomIntBounded => "Generate a random 32-bit unsigned integer less than `bound`.",
            RandomOp::NewRNGContext => {
                "Seed the RNG and return a new RNG context. Required before using other RNG ops, can be called only once."
            }
            RandomOp::DeleteRNGContext => "Discard the given RNG context.",
        }
        .to_string()
    }
}

impl MakeRegisteredOp for RandomOp {
    fn extension_id(&self) -> ExtensionId {
        EXTENSION_ID
    }

    fn extension_ref(&self) -> Weak<Extension> {
        Arc::downgrade(&EXTENSION)
    }
}

/// An extension trait for [Dataflow] providing methods to add
/// "tket.qsystem.random" operations.
pub trait RandomOpBuilder: Dataflow + UnwrapBuilder {
    /// Add a "tket.qsystem.random.random_int" op.
    fn add_random_int(&mut self, ctx: Wire) -> Result<[Wire; 2], BuildError> {
        Ok(self
            .add_dataflow_op(RandomOp::RandomInt, [ctx])?
            .outputs_arr())
    }

    /// Add a "tket.qsystem.random.random_float" op.
    fn add_random_float(&mut self, ctx: Wire) -> Result<[Wire; 2], BuildError> {
        Ok(self
            .add_dataflow_op(RandomOp::RandomFloat, [ctx])?
            .outputs_arr())
    }

    /// Add a "tket.qsystem.random.random_int_bounded" op.
    fn add_random_int_bounded(&mut self, ctx: Wire, bound: Wire) -> Result<[Wire; 2], BuildError> {
        Ok(self
            .add_dataflow_op(RandomOp::RandomIntBounded, [ctx, bound])?
            .outputs_arr())
    }

    /// Add a "tket.qsystem.random.new_rng_context" op.
    fn add_new_rng_context(&mut self, seed: Wire) -> Result<Wire, BuildError> {
        Ok(self
            .add_dataflow_op(RandomOp::NewRNGContext, [seed])?
            .out_wire(0))
    }

    /// Add a "tket.qsystem.random.delete_rng_context" op.
    fn add_delete_rng_context(&mut self, ctx: Wire) -> Result<(), BuildError> {
        self.add_dataflow_op(RandomOp::DeleteRNGContext, [ctx])?;
        Ok(())
    }
}

impl<D: Dataflow> RandomOpBuilder for D {}

#[cfg(test)]
mod test {
    use hugr::extension::simple_op::MakeExtensionOp;
    use hugr::ops::Value;
    use hugr::std_extensions::arithmetic::int_types::ConstInt;

    use hugr::builder::{DataflowHugr, FunctionBuilder};
    use hugr::HugrView;
    use strum::IntoEnumIterator;

    use super::*;

    #[test]
    fn create_extension() {
        assert_eq!(EXTENSION.name(), &EXTENSION_ID);

        for o in RandomOp::iter() {
            assert_eq!(
                RandomOp::from_def(EXTENSION.get_op(&o.op_id()).unwrap()),
                Ok(o)
            );
        }
    }

    #[test]
    fn test_random_op_builder() {
        let hugr = {
            let mut func_builder = FunctionBuilder::new(
                "random_op_builder",
                Signature::new(vec![], vec![int_type(5)]),
            )
            .unwrap();

            let seed =
                func_builder.add_load_const(Value::from(ConstInt::new_u(6, 123456).unwrap()));
            let maybe_ctx = func_builder.add_new_rng_context(seed).unwrap();
            let [ctx] = func_builder
                .build_unwrap_sum(
                    1,
                    option_type(RandomType::RNGContext.get_type(&Arc::downgrade(&EXTENSION))),
                    maybe_ctx,
                )
                .unwrap();
            let bound = func_builder.add_load_const(Value::from(ConstInt::new_u(5, 100).unwrap()));
            let [_, ctx] = func_builder.add_random_int_bounded(ctx, bound).unwrap();
            let [_, ctx] = func_builder.add_random_float(ctx).unwrap();
            let [rnd, ctx] = func_builder.add_random_int(ctx).unwrap();
            func_builder.add_delete_rng_context(ctx).unwrap();
            func_builder.finish_hugr_with_outputs([rnd]).unwrap()
        };
        hugr.validate().unwrap()
    }
}
