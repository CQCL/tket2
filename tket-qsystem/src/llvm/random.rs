//! LLVM lowering implementations for the "tket.qsystem.random" extension.

use tket::hugr::{self, llvm::inkwell};

use crate::extension::random::{self, RandomOp, CONTEXT_TYPE_NAME};
use anyhow::{anyhow, Result};
use hugr::llvm::custom::CodegenExtension;
use hugr::llvm::emit::func::EmitFuncContext;
use hugr::llvm::emit::EmitOpArgs;
use inkwell::builder::Builder;
use inkwell::context::Context;
use inkwell::types::{BasicTypeEnum, FloatType, IntType};
use inkwell::values::{BasicValueEnum, FunctionValue};
use tket::hugr::ops::ExtensionOp;
use tket::hugr::{HugrView, Node};

/// Codegen extension for `tket.qsystem.random` extension.
pub struct RandomCodegenExtension;

impl CodegenExtension for RandomCodegenExtension {
    fn add_extension<'a, H: HugrView<Node = Node> + 'a>(
        self,
        builder: tket::hugr::llvm::CodegenExtsBuilder<'a, H>,
    ) -> tket::hugr::llvm::CodegenExtsBuilder<'a, H>
    where
        Self: 'a,
    {
        builder
            .custom_type(
                (random::EXTENSION_ID, CONTEXT_TYPE_NAME.to_owned()),
                |session, _| Ok(rng_context_type(session.iw_context())),
            )
            .simple_extension_op(|context, args, op| RandomEmitter(context).emit(args, op))
    }
}

fn rng_context_type(context: &Context) -> BasicTypeEnum<'_> {
    context.struct_type(&[], false).into()
}

/// Lower the `tket.qsystem.random` extension.
struct RandomEmitter<'c, 'd, 'e, H: HugrView<Node = Node>>(&'d mut EmitFuncContext<'c, 'e, H>);

impl<'c, H: HugrView<Node = Node>> RandomEmitter<'c, '_, '_, H> {
    fn iw_context(&self) -> &'c Context {
        self.0.typing_session().iw_context()
    }

    fn i32_type(&self) -> IntType<'c> {
        self.iw_context().i32_type()
    }

    fn i64_type(&self) -> IntType<'c> {
        self.iw_context().i64_type()
    }

    fn float_type(&self) -> FloatType<'c> {
        self.iw_context().f64_type()
    }

    fn bool_type(&self) -> IntType<'c> {
        self.iw_context().bool_type()
    }

    fn rng_context(&self) -> BasicValueEnum<'c> {
        self.iw_context().const_struct(&[], false).into()
    }

    fn builder(&self) -> &Builder<'c> {
        self.0.builder()
    }

    /// Helper function to `emit` an RNG operation.
    fn emit_op(
        &self,
        args: EmitOpArgs<'c, '_, ExtensionOp, H>,
        name: &str,
        func: Result<FunctionValue<'c>>,
        input_indices: &[usize],
    ) -> Result<()> {
        let inputs: Vec<_> = input_indices.iter().map(|&i| args.inputs[i]).collect();
        let result = self
            .builder()
            .build_call(
                func?,
                &inputs.iter().map(|&v| v.into()).collect::<Vec<_>>(),
                name,
            )?
            .try_as_basic_value()
            .unwrap_left();
        args.outputs
            .finish(self.builder(), [result, self.rng_context()])
    }

    /// Function to help lower the `tket.qsystem.random` extension.
    fn emit(&self, args: EmitOpArgs<'c, '_, ExtensionOp, H>, op: RandomOp) -> Result<()> {
        match op {
            RandomOp::RandomInt => {
                let fn_random_int = self
                    .0
                    .get_extern_func("random_int", self.i32_type().fn_type(&[], false));
                self.emit_op(args, "rint", fn_random_int, &[])
            }
            RandomOp::RandomFloat => {
                let fn_random_float = self
                    .0
                    .get_extern_func("random_float", self.float_type().fn_type(&[], false));
                self.emit_op(args, "rfloat", fn_random_float, &[])
            }
            RandomOp::RandomIntBounded => {
                let fn_random_int_bounded = self.0.get_extern_func(
                    "random_rng",
                    self.i32_type().fn_type(&[self.i32_type().into()], false),
                );
                self.emit_op(args, "rintb", fn_random_int_bounded, &[1])
            }
            RandomOp::NewRNGContext => {
                let fn_random_seed = self.0.get_extern_func(
                    "random_seed",
                    self.iw_context()
                        .void_type()
                        .fn_type(&[self.i64_type().into()], false),
                )?;
                let [seed] = args
                    .inputs
                    .try_into()
                    .map_err(|_| anyhow!("NewRNGContext expects a seed argument"))?;
                self.builder()
                    .build_call(fn_random_seed, &[seed.into()], "rseed")?;
                args.outputs.finish(
                    self.builder(),
                    [self.bool_type().const_int(1, false).into()],
                )
            }
            RandomOp::DeleteRNGContext => args.outputs.finish(self.builder(), []),
        }
    }
}

#[cfg(test)]
mod test {
    use crate::extension::random::RandomOp;
    use hugr::extension::simple_op::MakeRegisteredOp;
    use hugr::llvm::check_emission;
    use hugr::llvm::test::llvm_ctx;
    use hugr::llvm::test::single_op_hugr;
    use hugr::llvm::test::TestContext;
    use rstest::rstest;

    use super::*;

    #[rstest]
    #[case::random_int(1, RandomOp::RandomInt)]
    #[case::random_float(2, RandomOp::RandomFloat)]
    #[case::random_int_bounded(3, RandomOp::RandomIntBounded)]
    #[case::new_rng_context(4, RandomOp::NewRNGContext)]
    #[case::delete_rng_context(5, RandomOp::DeleteRNGContext)]
    fn emit_random_codegen(
        #[case] _i: i32,
        #[with(_i)] mut llvm_ctx: TestContext,
        #[case] op: RandomOp,
    ) {
        llvm_ctx.add_extensions(|ceb| {
            ceb.add_extension(RandomCodegenExtension)
                .add_default_int_extensions()
                .add_float_extensions()
                .add_logic_extensions()
        });
        let ext_op = op.to_extension_op().unwrap().into();
        let hugr = single_op_hugr(ext_op);
        check_emission!(hugr, llvm_ctx);
    }
}
