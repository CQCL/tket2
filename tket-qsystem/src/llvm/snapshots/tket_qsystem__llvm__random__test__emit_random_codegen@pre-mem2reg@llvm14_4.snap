---
source: tket-qsystem/src/llvm/random.rs
expression: mod_str
---
; ModuleID = 'test_context'
source_filename = "test_context"

define private i1 @_hl.main.1(i64 %0) {
alloca_block:
  %"0" = alloca i1, align 1
  %"2_0" = alloca i64, align 8
  %"4_0" = alloca i1, align 1
  br label %entry_block

entry_block:                                      ; preds = %alloca_block
  store i64 %0, i64* %"2_0", align 4
  %"2_01" = load i64, i64* %"2_0", align 4
  call void @random_seed(i64 %"2_01")
  store i1 true, i1* %"4_0", align 1
  %"4_02" = load i1, i1* %"4_0", align 1
  store i1 %"4_02", i1* %"0", align 1
  %"03" = load i1, i1* %"0", align 1
  ret i1 %"03"
}

declare void @random_seed(i64)
